### ISSUE TITLE
Support Windows builds through clang

### ISSUE THREAD

**Issue by therc** [tensorflow/tensorflow#14425]
Right now, the tested configurations for Linux, Mac and Windows use gcc, clang and MSVC, as seen at 
https://www.tensorflow.org/install/install_sources#tested_source_configurations

Linux can be made to compile with clang, too, if you use the right magic trick (`--config=cuda_clang`). All that's left is Windows, which is probably the trickiest of the three.

Bonus points for allowing to cross-build for Windows under Linux. The main problem there might be with CUDA and getting its SDK installed on a Linux system (just a wild guess).

(Forked from an existing discussion under https://github.com/tensorflow/tensorflow/issues/12052)

**Comment by mrry** [comment]
This sounds potentially useful for development, but it's unlikely anyone on the team will take this on soon, because we already have a supported configuration that works well enough. Marking it as contributions welcome, in the hope that somebody wants to take it on.

**Comment by AtlantaPepsi** [comment]
sounds greatly interesting, I'll try to investigate into this.

**Comment by AtlantaPepsi** [comment]
I see that @gunan has written a windows clang toolchain under third_party/toolchains/preconfig/win/bazel_211, I'll probably assemble a new configuration with that and appropriate platform with new constraints. Right now I'm stuck on getting bazel working on my local machine, similar to #31608, where undname.exe couldn't be found under VC directory

**Comment by gunan** [comment]
That toolchain was constructed for MSVC 2019. AFAICR, all clang references to it are placeholders
@artem-b may be aware if we have windows clang toolchains checked in.

**Comment by Artem-B** [comment]
I don't think you need one. I've been using the default one provided by bazel.
I've basically followed these instructions:
https://docs.bazel.build/versions/master/windows.html#build-c-with-clang

**Comment by AtlantaPepsi** [comment]
@Artem-B Just that will do? I had seen those instructions, but was wondering if that's too simple for a "tricky" problem, then I realized this issue was opened back in 2017, perhaps Bazel didn't support this option back then.

But the platform target still need to be manually added by users afterwards right? I don't see it written anywhere in the source code. Do you think it would be still useful for me to add it somewhere and make a new config for clang on windows by incorporating those options and modifying the `config.py` ?

Another thing, MSVC seems to be the only option for CUDA compiler, I'm not sure how relatable is it to this issue or if it's already an open issue somewhere, but it might also be useful to include clang in CUDA configuration for windows?

**Comment by Artem-B** [comment]
BUILD/WORKSPACE files will need to have these modifications, though I think they will eventually make it into TF.
Adding it all to the config.py may be a bit premature -- people will run with it and will start complaining when things don't quite work, and builds on Windows do have a lot of sharp corners. 

As for CUDA compilation, clang should be able to do it in principle, but TF build needs a bit of work to put things together.
Basically the issue is that windows build with clang uses `clang-cl` which is a drop-in replacement for MSVC's `cl` and uses different command-line syntax. As far as bazel is concerned, CUDA compilation with clang is another C++ compilation with a few extra options. The problem is that TF currently only constructs these extra options to be used by regular `clang`  and will need to prefix all of them with `/clang` (I think) if they were to be used with `clang-cl` on windows.

> MSVC seems to be the only option for CUDA compiler, 

Do you mean NVCC? MSVC can't compile CUDA. That said, if you build with clang, your only choice is to whether enable CUDA or not. You can't (well, shouldn't, really) use NVCC in this case. I vaguely recall having to tweak config.py to force-enable CUDA for the build w/ clang.

**Comment by mohantym** [comment]
Hi @therc!
We are checking whether you are still looking help in this issue or not. You can say yes to clang support while running configure.py in after checking out r2.8/2.9 branch.

Attached relevant [thread ](https://www.tensorflow.org/install/source)from  source.
Thank you!

**Comment by mihaimaruseac** [comment]
CC @angerson @learning-to-play as there is some testing to use clang for building.

**Comment by learning-to-play** [comment]
@bani-intelaipg

**Comment by japm48** [comment]
According to the [release page](https://github.com/tensorflow/tensorflow/releases/tag/v2.16.1), clang is now used for the official TF builds of version 2.16.1.
The problem is there is no documentation (that I could find) on how to replicate that.

**Comment by learning-to-play** [comment]
@mraunak

**Comment by mraunak** [comment]
Hi @japm48,
First step!
Install LLVM
Go to the [LLVM downloads](https://github.com/llvm/llvm-project/releases/){:.external},
Download and install Windows-compatible LLVM in C:/Program Files/LLVM e.g., LLVM-17.0.6-win64.exe (recommended)

Note: To build with Clang on Windows, it is required to install both LLVM and Visual C++ Build tools as although Windows uses clang-cl.exe as the compiler, Visual C++ Build tools are needed to link to Visual C++ libraries
Environmental set up 
set BAZEL_SH=C:/msys64/usr/bin/bash.exe 
set BAZEL_VS=C:/Program Files/Microsoft Visual Studio/2022/BuildTools 
set BAZEL_VC=C:/Program Files/Microsoft Visual Studio/2022/BuildTools/VC 
set Bazel_LLVM=C:/Program Files/LLVM (explicitly tell Bazel where LLVM is installed by BAZEL_LLVM, needed while using CLANG)
set PATH=C:/Program Files/LLVM/bin;%PATH% (Optional, needed while using CLANG as Compiler)

Build with CLANG
Use --config=win_clang to build TenorFlow with the CLANG Compiler:

bazel build --config=win_clang --repo_env=TF_PYTHON_VERSION=3.10 //tensorflow/tools/pip_package:wheel --repo_env=WHEEL_NAME=tensorflow_cpu

**Comment by japm48** [comment]
Thanks @mraunak !

I got this error: `no such target '//tensorflow/tools/pip_package:wheel'`.

So, I tried with `//tensorflow/tools/pip_package:build_pip_package`.

Also, the `build_pip_package` executable requires rsync on msys2 to work.

Apart from that, `build_pip_package` complained about `invalid command 'bdist_wheel'`, even if `wheel` package was installed. After some research I found that it uses the python executable found in `bazel-tensorflow/external`, so installing `wheel` in that did the trick.

----

For the record, the complete setup is as follows (all commands executed in an "as admin" `cmd.exe` prompt):
 - VS2022 Tools: current latest version.
 - MSYS2 installed at `C:\msys64` updated and with required packages installed:
   - `pacman -Syu` (requires a console restart)
   - `pacman -S git patch unzip rsync`
 - Python 3.10 installed at `C:\Python310`  (3.11 fails with error `invalid command 'bdist_wheel'`)
   - `C:\Python310\python.exe -m pip install -U pip` (to update pip itself)
   - `C:\Python310\python.exe -m pip install -U six numpy wheel packaging`
   - `C:\Python310\python.exe -m pip install -U keras_preprocessing --no-deps`
 - bazelisk installed and reachable in PATH (e.g. using `choco install bazelisk`)
 - LLVM version 17.0.6, as recommended.
 - Tensorflow 2.16.1 downloaded from the [releases page](https://github.com/tensorflow/tensorflow/releases/tag/v2.16.1).
   - Folder containing the sources must be named `tensorflow`, and not `tensorflow-2.16.1` (otherwise `build_pip_package` fails).

```
set BAZEL_SH=C:/msys64/usr/bin/bash.exe
set BAZEL_VS=C:/Program Files/Microsoft Visual Studio/2022/BuildTools
set BAZEL_VC=C:/Program Files/Microsoft Visual Studio/2022/BuildTools/VC 
set Bazel_LLVM=C:/Program Files/LLVM
set PATH=C:/Python310/;C:/Python310/Scripts;C:/Program Files/LLVM/bin;%PATH%

bazel build --config=win_clang --repo_env=TF_PYTHON_VERSION=3.10 //tensorflow/tools/pip_package:build_pip_package --repo_env=WHEEL_NAME=tensorflow_cpu
```

And finally:
```
.\bazel-tensorflow\external\python_x86_64-pc-windows-msvc\python.exe -m pip install wheel
.\bazel-bin\tensorflow\tools\pip_package\build_pip_package ./tensorflow_pkg
```

----

I would say this should be documented in the build instructions page before closing this issue (you can copy any of the above if necessary).

**Comment by mraunak** [comment]
Hi @japm48, Thank you very much! I will incorporate the changes suggested by you. We recently made a changed the target to build the TensorFlow(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/pip_package/BUILD#L265)
Request you to clone to the latest TensorFlow Repo and try to build the target //tensorflow/tools/pip_package:wheel

**Comment by japm48** [comment]
@mraunak I will try again with the latest dev version when I have the time and resources, and will report back. The issue with the missing wheel dependency is perhaps fixed in that version.

Just a quick question: given the presence of the new "hermetic Python", is it no longer necessary to manually install Python itself on the build system?

**Comment by mraunak** [comment]
yes, the TensorFlow build should compile without asking for Python installation if --repo_env=TF_PYTHON_VERSION.. is given

**Comment by japm48** [comment]
Forgot this existed, sorry. I'll have a go in the next few days to confirm if this is enough.

**Comment by mihaimaruseac** [comment]
All builds are now being done via clang, so this is likely resolved?

**Comment by japm48** [comment]
Hi @mihaimaruseac , @mraunak .

(Let me know if I should repost this somewhere else, perhaps here?: https://groups.google.com/a/tensorflow.org/g/build)

I wrote a bunch of PowerShell scripts to fully automate the whole process including the dependency install (with winget):
  * https://github.com/japm48/tf-build-windows/
  * Feel free to copy it, 0BSD-licensed (I can change the license if you need that).
  * Check `_config.ps1` for the parameters available to tweak.
  * Scripts are supposed to be executed in order 01 to 04, with powershell as admin (need `ExecutionPolicy Bypass` or similar, see first comment for an example)

----
Some notes:
* I would recommend starting from a clean Windows setup, given the fragility of the whole process.
  * A VM works fine, but perhaps using Windows Sandbox is a better option (haven't tried this yet).
* I got a Clang compiler crash with some avx512 code, this was automagically solved by upgrading to LLVM 18.1.8 (original version was 17.0.6).
* Python 3.12 works just fine; only thing needed is to set `--repo_env=TF_PYTHON_VERSION=3.12` on `bazel build`.
* Bazel likes to constantly create and delete a bunch of 10s-MB-sized files in `%TMP%`, which is not very Windows/NTFS-friendly. I would recommend setting $TMP_DIR into a ramdisk drive.
  * If lots of RAM are available (around 30 GB required to reserve), I would suggest also setting $BUILD_DIR to a ramdisk drive dir.
  * Recommended option: OSFMount. ImDisk did not work for me, probably something related to symlink compatibility.
* The sources directory still needs to be called "tensorflow", otherwise an obscure Python error is triggered.
* One particular Bazel subtarget that is very demanding is `//tensorflow/core/kernels:cwise_op`, compiling it with all CPU cores can easily exhaust the available RAM (each compiler process takes around 4 GB, with a maximum of 7 GB)
  * I recommend compiling it separately first with a reduced number of CPUs (this is also done in my script).
* In a 42-core VM, the build took around 7,000 seconds in total (6,000s correspond to the `cwise_op` module).
  * This required around *160 GB* peak RAM usage (including ramdisk).
* Standard Windows chores (not currently covered in the scripts):
  * automatic suspend should be disabled
  * relevant folders should be set to be ignored by Windows Defender

---

With all that, everything works fine finally. From my side it is now fine to close the issue.

**Comment by mihaimaruseac** [comment]
This is amazing work! Thank you for detailing all the steps, the hurdles and the time+memory requirements

**Comment by Kayyuri** [comment]
@japm48, Thanks for the detailed explanation!

Hi @therc, Since this issue has been resolved. Please feel free to close this issue if you have no further concerns.
